### PR TITLE
Personal/nitinsingla/rebase main

### PR DESCRIPTION
**Pull request by nitin-deamon**
Write changes with rebase to main.

### CODE REVIEW

File: nfs/src/rpc_task.cpp [hunk 1]
```diff
@@ -184,6 +207,180 @@ static void lookup_callback(
     }
 }
 
+#if 0
+static void flush_callback(
+    struct rpc_context* /* rpc */,
+    int rpc_status,
+    void *data,
+    void *private_data)
+{
```
Comment (linuxsmiths) [inline nfs/src/rpc_task.cpp]: why have you kept it commented?
other than the name difference is it different from write_flush_callback()?
were you originally thinking of keeping a separate callback for flush and write and then you ended up using one?

File: nfs/src/rpc_task.cpp [hunk 2]
```diff
@@ -304,6 +501,294 @@ void rpc_task::run_lookup()
...
+        args.count  = membuf->length;
+        args.stable = FILE_SYNC;
+        args.data.data_len = membuf->length;
+        args.data.data_val = (char *) membuf->buffer;
+
+        // Allocate new rpc task.
+        auto flush_task = client->get_rpc_task_helper()->alloc_rpc_task();
+        flush_task->optype = FUSE_FLUSH;
+
+        struct write_flush_context *cb_context = new write_flush_context(chunk, flush_task, file_ino);
```
Comment (linuxsmiths) [inline nfs/src/rpc_task.cpp]: why do you need to create an rpc task here? 
see related comment in run_flush()

File: nfs/src/rpc_task.cpp [hunk 3]
```diff
@@ -304,6 +501,294 @@ void rpc_task::run_lookup()
...
+        /*
+         * As of now, whenever write comes we get byte_chunk for corresponding
+         * offset and length. Copy the buffer from fuse to that, mark the buffer dirty.
+         * After that we issue write rpc to blob and complete the write command.
+         * Ideally all the chunks marked dirty in chunkmap, already issued to blob.
+         * We just need to wait for them to complete. That's what flush will do.
+         *
+         * Note: There is race condition to that, where we marked the buffer dirty, before
+         *       flushing it to blob, flush command come.
+         *
```
Comment (linuxsmiths) [inline nfs/src/rpc_task.cpp]: we should not have to worry about these.
I suggest write a function called sync_membuf(). It takes a membuf parameter and ensures it's sync'ed to the backing blob. Right now you can write it here, later we can decide if that moves to the membuf class itself.
Once you have this function, then write()/flush()/sync() just call this function on all dirty membufs. write() gets dirty membufs from copy_to_cache() and flush/sync will get it using get_dirty_bc().
Then we don't need to worry about flush and write racing. We just need to make sure a dirty membuf is sync'ed.
Whoever gets to it firsy, syncs it, the other one doesn't do anything.

File: nfs/src/rpc_task.cpp [hunk 4]
```diff
@@ -304,6 +501,294 @@ void rpc_task::run_lookup()
...
+            assert(membuf->buffer != nullptr);
+            assert(membuf->length != 0);
+
+            if (nfs_inode->get_write_error() == 0)
+            {
+                AZLogDebug("membuf dirty and no error set offset {}, Length {}", membuf->offset, membuf->length);
+                assert(0);
+            }
+        #if 0
+            WRITE3args args;
```
Comment (linuxsmiths) [inline nfs/src/rpc_task.cpp]: again, did u intend to use it?

File: nfs/src/rpc_task.cpp [hunk 5]
```diff
@@ -304,6 +501,294 @@ void rpc_task::run_lookup()
...
+            // Read from the fd and copy to buffer.
+            if (fd_set) {
+                auto len = chunk.length;
+                auto dst_off = 0;
+
+                // Let's read in a loop.
+                while(len) {    
+                    if (seek_set) {
+                        res = pread(bufv->buf[bufv->idx].fd, chunk.get_buffer() + dst_off, len, src_off);
+                    } else {
```
Comment (linuxsmiths) [inline nfs/src/rpc_task.cpp]: I think the fd mode is provided for splicing. Unless we use splice() it's not beneficial to use the fd mode. We can as well turn it off by not setting FUSE_CAP_SPLICE_READ. Did you try that?

File: nfs/src/rpc_task.cpp (hunk 1 above)
Comment (nitin-deamon) [reply nfs/src/rpc_task.cpp]: Removed the commented code.

File: nfs/src/rpc_task.cpp (hunk 2 above)
Comment (nitin-deamon) [reply nfs/src/rpc_task.cpp]: As discussed to put back pressure on incoming writes, created new rpc.

File: nfs/src/rpc_task.cpp (hunk 3 above)
Comment (nitin-deamon) [reply nfs/src/rpc_task.cpp]: Done

File: nfs/src/rpc_task.cpp (hunk 4 above)
Comment (nitin-deamon) [reply nfs/src/rpc_task.cpp]: removed the commented code.

File: nfs/src/rpc_task.cpp (hunk 5 above)
Comment (nitin-deamon) [reply nfs/src/rpc_task.cpp]: FUSE_CAP_SPLICE_READ worked, when I was implementing, tried setting FUSE_CAP_SPLICE_WRITE off which didn't work.

**Review by linuxsmiths — APPROVED** [review]
approving w/o review, will review/audit after merging.

File: nfs/src/rpc_task.cpp [hunk 6]
```diff
@@ -318,6 +435,206 @@ void rpc_task::run_lookup()
...
+    bool rpc_retry = false;
+    auto membuf = chunk.get_membuf();
+
+    // Verify the membuf.
+    assert(membuf != nullptr);
+    assert(membuf->is_inuse());
+
+    // Get the membuf which need to flush.
+    membuf->set_locked();
+    if (membuf->is_dirty() && !membuf->is_flushing())
```
Comment (linuxsmiths) [inline nfs/src/rpc_task.cpp:523]: is_flushing check must be before the set_locked(), o/w how do we get the benefit of not having to wait for lock if someone is already fliushing.

File: nfs/src/rpc_task.cpp [hunk 7]
```diff
@@ -318,6 +435,206 @@ void rpc_task::run_lookup()
...
+    {
+        auto membuf = chunk.get_membuf();
+        assert(membuf != nullptr);
+
+        membuf->set_locked();
+        assert(membuf->is_inuse());
+
+        if (membuf->is_dirty()) {
+            assert(nfs_inode->get_write_error() != 0);
+            AZLogDebug ("Writes failed with error: {}", nfs_inode->get_write_error());
```
Comment (linuxsmiths) [inline nfs/src/rpc_task.cpp:628]: technically some other thread can get the lock before we get the lock here and it may set it to dirty.

**Pull request merged**